packages/core/contracts/zodiac/OptimisticOracleModule.sol

```diff
@@ -0,0 +1,271 @@
+        __Ownable_init();
+        (
+            address _finder,
+            address _owner,
+            address _collateral,
+            uint256 _bond,
+            string memory _rules,
+            bytes32 _identifier,
+            uint64 _liveness
+        ) = abi.decode(initializeParams, (address, address, address, uint256, string, bytes32, uint64));
+        finder = FinderInterface(_finder);
+        require(_getCollateralWhitelist().isOnWhitelist(address(_collateral)), "bond token not supported");
+        collateral = IERC20(_collateral);
+        bond = _bond;
+        rules = _rules;
+        identifier = _identifier;
+        require(_liveness > 0, "liveness can't be 0");
+        liveness = _liveness;
+        setAvatar(_owner);
+        setTarget(_owner);
+        transferOwnership(_owner);
+        _sync();
+
+        emit OptimisticOracleModuleDeployed(_owner, avatar, target);
+    }
+
+    function setBond(uint256 _bond) public onlyOwner {
+        // Value of the bond required for proposals, in addition to the final fee.
+        bond = _bond;
+    }
+
+    function setCollateral(IERC20 _collateral) public onlyOwner nonReentrant {
+        // ERC20 token to be used as collateral (must be approved by UMA Store contract).
+        require(_getCollateralWhitelist().isOnWhitelist(address(_collateral)), "bond token not supported");
+        collateral = _collateral;
+    }
+
+    function setRules(string memory _rules) public onlyOwner {
+        // Set reference to the rules for the avatar (e.g. an IPFS hash or URI).
+        rules = _rules;
+    }
+
+    function setLiveness(uint64 _liveness) public onlyOwner {
+        // Set liveness for disputing proposed transactions.
+        require(_liveness > 0, "liveness can't be 0");
+        liveness = _liveness;
+    }
+
+    function setIdentifier(bytes32 _identifier) public onlyOwner {
+        // Set identifier which is used along with the rules to determine if transactions are valid.
+        identifier = _identifier;
+    }
+
+    function sync() public nonReentrant {
+        // Sync the store and oracle contract addresses as well as the final fee.
+        _sync();
+    }
+
+    function proposeTransactions(Transaction[] memory _transactions, bytes memory _explanation) public nonReentrant {
+        // Create a proposal with a bundle of transactions.
+        // note: Based in part on the UMA Governor contract.
+        // https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/oracle/implementation/Governor.sol
+        // note: Pptional explanation explains the intent of the transactions to make comprehension easier.
+        uint256 id = proposals.length;
+        uint256 time = getCurrentTime();
+        address proposer = msg.sender;
+
+        // Add a zero-initialized element to the proposals array.
+        Proposal storage proposal = proposals.push();
+        proposal.requestTime = time;
+        proposal.ancillaryData = _explanation;
```

Comment:
That's a good idea.

Reply:
However, I believe that would limit the `_explanation` to a `bytes32` length since the `AncillaryData` library has an `appendKeyValueBytes32` function but no way to append a key-value with arbitrary length bytes.

Reply:
I'll include that warning in a UMIP update.